### PR TITLE
feat: handle upload zip with directories

### LINKED ISSUE #452: Allow uploading zip files consisting of exactly one top-level directory containing the actual files
Screen Shot 2023-06-19 at 05 57 39

Currently this only works when the zip file consists of files at the top-level.

But sometimes, when zipping, one zips the parent directory, not files. So when Judgels unzips the file, it only finds one top-level entry, which is a directory, and skips it.

We need to support this case by simply recursing to the (only) directory when iterating the files.

If there are multiple top-level directories, maybe we can ignore them at all (treat this as invalid zip file, like existing behavior)

### PR DESCRIPTION
**Pull request by kadeksuryam**
Resolve #452

### CODE REVIEW

**Comment by kadeksuryam** [comment]
Hello. My approach for this issue is that every file that exists in a directory will be moved to the outermost parent directory, which is the resource directory. For example, given this zip file with these files and directories:
```
dir1
  file1
  dir2
    file2
dir3
  file3
file4
```
All of the files: file1, file2, file3, and file4, will be successfully unzipped to the resource folder without the dir1, dir2, and dir3 directories. In case of duplicate filenames in different directories, the last processed file will be selected.

What do you think? @fushar

File: judgels-backends/judgels-commons/judgels-fs/src/main/java/judgels/fs/local/LocalFileSystem.java [hunk 1]
```diff
@@ -106,21 +107,32 @@ public String getPrivateFileUrl(Path filePath) {
...
     @Override
-    public void uploadZippedFiles(Path dirPath, InputStream content, boolean includeDirectory) {
+    public void uploadZippedFiles(Path dirPath, InputStream content) {
         try {
             File destDir = baseDir.resolve(dirPath).toFile();
             byte[] buffer = new byte[4096];
             int entries = 0;
             long total = 0;
+
+            Set<String> directories = Sets.newHashSet();
```
Comment (fushar) [inline judgels-backends/judgels-commons/judgels-fs/src/main/java/judgels/fs/local/LocalFileSystem.java]: I think this whole logic can be simplified by always taking `FilenameUtils.getName(ze.getName())` as the filename for all non-directory entries. Something like (pseudocode):

```
if (ze.isDirectory()) {
  continue;
}

String filename = FilenameUtils.getName(ze.getName());
File file = new File(destDir, filename);

// etc
```

What do you think?

**Review by fushar — CHANGES_REQUESTED** [review]
Thanks for your contribution!

Unfortunately, this is a bit complicated. We have to ignore certain directories:

https://github.com/ia-toki/judgels/blob/314fe62f1ea69bf4bce01ec0bedb8a30b5e462b8/judgels-backends/judgels-commons/judgels-fs/src/main/java/judgels/fs/local/LocalFileSystem.java#L33-L36

- `.gitkeep`
- `__MACOSX`: this is present when we compress the ZIP on Mac. See:

Screenshot 2023-10-21 at 09 07 04

So we need to ignore a file if it lives in any of the ignorable directories. Something like this:

```
String filename = FilenameUtils.getName(ze.getName());
if (IGNORABLE_FILES.stream().anyMatch(dir -> filename.contains(dir + "/"))) {
  // skip this file
}
```

---

For reference, here is the result of uploading zip files from Mac, with the current PR, which is wrong

image

**Review by kadeksuryam — COMMENTED** [review]
handle file entry in ignorable dir

**Review by fushar — APPROVED** [review]

**Pull request merged**